torch/onnx/_internal/exporter.py

```diff
@@ -155,21 +156,215 @@ def serialize(
+        ...
+
+
+class InputFormatter:
+    """A class that formats the PyTorch model inputs to exported ONNX model inputs format."""
+
+    _input_format_steps: List[InputFormatStep]
+
+    def __init__(self, input_format_steps: Optional[List[InputFormatStep]] = None):
+        self._input_format_steps = input_format_steps or []
+
+    @_beartype.beartype
+    def append_step(self, step: InputFormatStep) -> None:
+        """Appends a step to the input format steps.
+
+        Args:
+            step: The step to append.
+        """
+        self._input_format_steps.append(step)
+
+    @_beartype.beartype
+    def to_onnx(self, *model_args, **model_kwargs) -> Sequence[torch.Tensor]:
+        """Converts the PyTorch model inputs to exported ONNX model inputs format.
+
+        Args:
+            model_args: The PyTorch model inputs.
+            model_kwargs: The PyTorch model keyword inputs.
+
+        Returns:
+            A sequence of tensors converted from PyTorch model inputs.
+        """
+        args: Sequence[Any] = model_args
+        kwargs: Mapping[str, Any] = model_kwargs
+        for step in self._input_format_steps:
+            args, kwargs = step.format(args, kwargs)
+        assert not kwargs
+        return args
+
+
+@runtime_checkable
+class OutputFormatStep(Protocol):
+    def format(self, model_outputs: Any) -> Any:
+        ...
+
+
+class OutputFormatter:
+    """A class that formats the PyTorch model outputs to exported ONNX model outputs format."""
+
+    _output_format_steps: List[OutputFormatStep]
+
+    def __init__(self, output_format_steps: Optional[List[OutputFormatStep]] = None):
+        self._output_format_steps = output_format_steps or []
+
+    @_beartype.beartype
+    def append_step(self, step: OutputFormatStep) -> None:
+        """Appends a step to the output format steps.
+
+        Args:
+            step: The step to append.
+        """
+        self._output_format_steps.append(step)
+
+    @_beartype.beartype
+    def to_onnx(self, model_outputs: Any) -> Sequence[torch.Tensor]:
+        """Converts the PyTorch model outputs to exported ONNX model outputs format.
+
+        Args:
+            model_outputs: The PyTorch model outputs.
+
+        Returns:
+            PyTorch model outputs in exported ONNX model outputs format.
+        """
+        for step in self._output_format_steps:
+            model_outputs = step.format(model_outputs)
+        return model_outputs
+
+
 class ExportOutput:
     """An in-memory representation of a PyTorch model that has been exported to ONNX."""
 
     _model_proto: Final[onnx.ModelProto]
+    _input_formatter: Final[InputFormatter]
+    _output_formatter: Final[OutputFormatter]
 
     @_beartype.beartype
-    def __init__(self, model_proto: onnx.ModelProto):
+    def __init__(
+        self,
+        model_proto: onnx.ModelProto,
+        input_formatter: InputFormatter,
+        output_formatter: OutputFormatter,
```

Comment:
> The goal of `ExportOutput` is to hold the exported model.

I think anything specifically associated with a particular export run is reasonable to be counted as "export output". E.g., diagnostic log, formatter. It is pointed out in the other reply that formatter is not a static class/object, but something created by the export call, as is model_proto.

Reply:
I am still unsure about it, it really looks a helper function that doesn't need to be created by the exporter.

The `diagnostic` and `model_proto` properties are model+data specific states, unique to that export session execution. One cannot achieve their content through reverse engineering it.

`formatter`, on the other hand, although looks unique to the input data/fx graph extractor used during export, one could easily implement a function without using the exporter to achieve the same thing. It is kind of data conversion recipe that is 1) used by exporter and 2) might be useful for CI/niche users.

For 1), we don't need to expose it publicly. For 2) it is not the most common path, so arguably maybe not too needed by a public API. Especially because there are ways to go around it.

For example,  one has the model `data` (which does not belong to the exporter) and knows which FX graph extractor they want to use, say, `torch.fx.symbolic_trace`, they know how to modify `data` to successfully create the FX graph. `torch.onnx.dynamo_export` and `ExportOutput` are not needed at all. Keeping internalized, decreases public surface, things gets more decoupled and easy to modify later.